storm-core/src/jvm/org/apache/storm/scheduler/IsolationScheduler.java

```diff
@@ -0,0 +1,417 @@
+                failedTopologyIds.add(topologyId);
+            }
+            List<Integer> workerNum = distributionSortedAmts(topologyMachineDistributions.get(topologyId));
+            for (Integer num : workerNum) {
+                HostAssignableSlots hostSlots = hss.peek();
+                List<WorkerSlot> slot = hostSlots != null ? hostSlots.getWorkerSlots() : null;
+
+                if (slot != null && slot.size() >= num.intValue()) {
+                    hss.poll();
+                    cluster.freeSlots(hostUsedSlots.get(hostSlots.getHostName()));
+                    for (WorkerSlot tmpSlot : slot.subList(0, num)) {
+                        Set<ExecutorDetails> executor = removeElemFromSet(executorSet);
+                        cluster.assign(tmpSlot, topologyId, executor);
+                    }
+                    cluster.blacklistHost(hostSlots.getHostName());
+                }
+            }
+        }
+
+        if (failedTopologyIds.size() > 0) {
+            LOG.warn("Unable to isolate topologies " + failedTopologyIds
+                    + ". No machine had enough worker slots to run the remaining workers for these topologies. "
+                    + "Clearing all other resources and will wait for enough resources for "
+                    + "isolated topologies before allocating any other resources.");
+            // clear workers off all hosts that are not blacklisted
+            Map<String, Set<WorkerSlot>> usedSlots = hostUsedSlots(cluster);
+            Set<Map.Entry<String, Set<WorkerSlot>>> entries = usedSlots.entrySet();
+            for (Map.Entry<String, Set<WorkerSlot>> entry : entries) {
+                if (!cluster.isBlacklistedHost(entry.getKey())) {
+                    cluster.freeSlots(entry.getValue());
+                }
+            }
+        } else {
+            // run default scheduler on non-isolated topologies
+            Set<String> allocatedTopologies = allocatedTopologies(topologyWorkerSpecs);
+            Topologies leftOverTopologies = leftoverTopologies(topologies, allocatedTopologies);
+            DefaultScheduler.defaultSchedule(leftOverTopologies, cluster);
+        }
+        cluster.setBlacklistedHosts(origBlacklist);
+    }
+
+    public Set<ExecutorDetails> removeElemFromSet(Set<Set<ExecutorDetails>> executorsSets) {
+        Set<ExecutorDetails> elem = executorsSets.iterator().next();
+        executorsSets.remove(elem);
+        return elem;
+    }
+
+    private List<TopologyDetails> isolatedTopologies(Collection<TopologyDetails> topologies) {
+        Set<String> topologyNames = isoMachines.keySet();
+        List<TopologyDetails> isoTopologies = new ArrayList<TopologyDetails>();
+        for (TopologyDetails topo : topologies) {
+            if (topologyNames.contains(topo.getName())) {
+                isoTopologies.add(topo);
+            }
+        }
+        return isoTopologies;
+    }
+
+    private Set<String> isolatedTopoplogyIds(List<TopologyDetails> topologies) {
```

Comment:
I'd rather rename to `extractTopologyIds` since method logic is general, not related to isolation.